File: src/loki_code/tests/fixtures/sample_code/test_sample.ts
/**
 * Sample TypeScript code for testing Tree-sitter analysis.
 */

interface CalculatorOptions {
    precision?: number;
    verbose?: boolean;
}

interface CalculationResult {
    result: number;
    operation: string;
    timestamp: Date;
}

class Calculator {
    private history: CalculationResult[] = [];
    private options: CalculatorOptions;
    
    constructor(options: CalculatorOptions = {}) {
        this.options = { precision: 2, verbose: false, ...options };
    }
    
    public add(a: number, b: number): number {
        const result = a + b;
        const calculation: CalculationResult = {
            result,
            operation: `${a} + ${b}`,
            timestamp: new Date()
        };
        this.history.push(calculation);
        return result;
    }
    
    public multiply(a: number, b: number): number {
        const result = a * b;
        const calculation: CalculationResult = {
            result,
            operation: `${a} * ${b}`,
            timestamp: new Date()
        };
        this.history.push(calculation);
        return result;
    }
    
    public getHistory(): CalculationResult[] {
        return [...this.history];
    }
}

function fibonacci(n: number): number {
    if (n <= 1) return n;
    return fibonacci(n - 1) + fibonacci(n - 2);
}

type DataItem = string | number | boolean | null;
type DataObject = Record<string, DataItem>;

interface ProcessingOptions {
    verbose?: boolean;
    timeout?: number;
}

interface ProcessingResult {
    status: "processing" | "completed" | "error";
    inputSize: number;
    timestamp: string;
    [key: string]: any;
}

function complexFunction(data: DataObject, options: ProcessingOptions = {}): ProcessingResult {
    const defaultOptions: ProcessingOptions = { verbose: false, timeout: 30 };
    const config = { ...defaultOptions, ...options };
    
    const result: ProcessingResult = {
        status: "processing",
        inputSize: Object.keys(data).length,
        timestamp: new Date().toISOString()
    };
    
    // Process data with type checking
    for (const [key, value] of Object.entries(data)) {
        if (typeof value === 'string') {
            result[key] = value.toUpperCase();
        } else if (typeof value === 'number') {
            result[key] = value * 2;
        } else if (typeof value === 'boolean') {
            result[key] = !value;
        } else {
            result[key] = String(value);
        }
    }
    
    result.status = "completed";
    return result;
}

// Generic function with constraints
function processItems<T extends { id: number }>(items: T[]): T[] {
    return items.filter(item => item.id > 0);
}

// Async function with proper typing
async function fetchData(url: string): Promise<{ data: any; status: number }> {
    // Simulated async operation
    return new Promise(resolve => {
        setTimeout(() => {
            resolve({ data: { message: "success" }, status: 200 });
        }, 100);
    });
}

export {
    Calculator,
    fibonacci,
    complexFunction,
    processItems,
    fetchData,
    type CalculatorOptions,
    type ProcessingResult
};

File: src/loki_code/tests/fixtures/sample_code/test_sample.rs
/**
 * Sample Rust code for testing Tree-sitter analysis.
 */

use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct Calculator {
    history: Vec<String>,
}

impl Calculator {
    pub fn new() -> Self {
        Calculator {
            history: Vec::new(),
        }
    }
    
    pub fn add(&mut self, a: i32, b: i32) -> i32 {
        let result = a + b;
        self.history.push(format!("{} + {} = {}", a, b, result));
        result
    }
    
    pub fn multiply(&mut self, a: i32, b: i32) -> i32 {
        let result = a * b;
        self.history.push(format!("{} * {} = {}", a, b, result));
        result
    }
    
    pub fn get_history(&self) -> &Vec<String> {
        &self.history
    }
}

pub fn fibonacci(n: u32) -> u32 {
    match n {
        0 => 0,
        1 => 1,
        _ => fibonacci(n - 1) + fibonacci(n - 2),
    }
}

#[derive(Debug)]
pub struct ProcessingOptions {
    pub verbose: bool,
    pub timeout: u32,
}

impl Default for ProcessingOptions {
    fn default() -> Self {
        ProcessingOptions {
            verbose: false,
            timeout: 30,
        }
    }
}

pub fn complex_function(
    data: HashMap<String, String>, 
    options: Option<ProcessingOptions>
) -> HashMap<String, String> {
    let opts = options.unwrap_or_default();
    let mut result = HashMap::new();
    
    result.insert("status".to_string(), "processing".to_string());
    result.insert("input_size".to_string(), data.len().to_string());
    
    // Process data
    for (key, value) in data.iter() {
        result.insert(key.clone(), value.to_uppercase());
    }
    
    result.insert("status".to_string(), "completed".to_string());
    result
}

// Generic function with traits
pub fn process_items<T: Clone + std::fmt::Debug>(items: Vec<T>) -> Vec<T> {
    items.into_iter().collect()
}

// Error handling with Result
#[derive(Debug)]
pub enum CalculationError {
    DivisionByZero,
    InvalidInput,
}

impl std::fmt::Display for CalculationError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            CalculationError::DivisionByZero => write!(f, "Division by zero"),
            CalculationError::InvalidInput => write!(f, "Invalid input"),
        }
    }
}

impl std::error::Error for CalculationError {}

pub fn safe_divide(a: f64, b: f64) -> Result<f64, CalculationError> {
    if b == 0.0 {
        Err(CalculationError::DivisionByZero)
    } else {
        Ok(a / b)
    }
}

fn main() {
    let mut calc = Calculator::new();
    println!("5 + 3 = {}", calc.add(5, 3));
    println!("4 * 7 = {}", calc.multiply(4, 7));
    
    match safe_divide(10.0, 2.0) {
        Ok(result) => println!("10 / 2 = {}", result),
        Err(e) => println!("Error: {}", e),
    }
}